core/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -2180,7 +2184,15 @@ private EngineConfig newEngineConfig(EngineConfig.OpenMode openMode) {
                 forceNewHistoryUUID = true;
                 break;
             default:
-                throw new AssertionError("unknown recovery type: [" + shardRouting.recoverySource().getType() + "]");
+                throw new AssertionError("unknown recovery type: [" + recoveryType + "]");
+        }
+        final IndexCommit startingCommit;
+        if (recoveryType == RecoverySource.Type.EXISTING_STORE) {
```

Comment:
why does this needs to be out of engine? can't we do it in the engine construct as an invariant when opening an index (and translog)

Reply:
Yep, I will it to the Engine.